### DISCUSSION TITLE
Is there some way to not use the shorthand notation L(k,n) in the program.

### DISCUSSION THREAD

**Discussion by Unaimend** [hklarner/detection_of_markers_for_discrete_phenotypes#5] [Q&A] [answered]
See title. I would like to see the whole factorization sets if possible. Is there some easy way of doing this

**Comment by hklarner** [comment]
Can you send me a marker file whose factorization contains unwanted shorthand notation? (via email or here in this issue)

**Reply by Unaimend** [reply]
Sorry, I was a little unclear. I am currently looking into the evolution of marker sets if we "refine" the phenotype. Imagine we have three phenotypes M, H, E. I  am interested in what happens to the marker sets when we now, for example, split the Phenotype H into H_1, ..., H_n. 

![image](https://github.com/hklarner/detection_of_markers_for_discrete_phenotypes/assets/7808379/526062c0-cc8b-4341-a6d6-49595c31c951)

During `markers-factorize` you provide the markers sets with the L(n,k) shorthand notation in the terminal.  I would like to know if there is a way to expand this notation. Ie get all 27 sets of size 2 printed

**Reply by hklarner** [reply]
Yes, I understand. I could add an option to `markers-factorize` that exports all L factors in a csv, for example. But I need an example that produces long factors to do the programming. I could not come up with one myself, I have tried a few randomly. Can you send the parameters for `problem-create` and `problem-solve` that result in some L terms? Preferably using the Bnet files in the repo. Or you share your own Bnet file.

**Comment by Unaimend** [comment]
Here is the bnet, I used one from the pyboolnet repo. Suffix changed to .txt because GitHub complained. If you prefer I can also try to make a PR.

[calzone_example.bnet.txt](https://github.com/hklarner/detection_of_markers_for_discrete_phenotypes/files/15481120/calzone_example.bnet.txt)

Or the link to pybool net
https://github.com/hklarner/pyboolnet/tree/master/pyboolnet/repository/calzone_cellfate

**Reply by hklarner** [reply]
I don't know which phenotypes generate L factors. Here is what I tried:
```
$ biomarkers problem-create --bnet calzone_example.bnet --phenotype TNF=0,SMAC=1
created json file: fname=tmp_problem.json
n_steady_states: 27
component_names: {20: 'SMAC', 22: 'TNF'}
phenotype_subspace: {22: 0, 20: 1}
   phenotype_index           phenotype  n_steady_states
0                0  not {22: 0, 20: 1}               19
1                1      {22: 0, 20: 1}                8

$ biomarkers problem-solve
created tmp_program.asp
clingo --models=0 --opt-mode=optN --enum-mode=domRec --heuristic=Domain --dom-mod=5,16
cpu time: 0:00:00
n_marker_sets=46
first 3 marker sets: [[4, 19, 23], [1, 19, 23], [19, 23, 25]]
created json file: fname=tmp_markers.json

$ biomarkers markers-factorize
 n_components  n_markers                            factorization  n_factors_used  n_factors_available
            3         40 S(22,23) * S(0,13,15,19) * S(1,2,4,5,25)               3                    3
            2          6                    S(22,23) * S(6,12,20)               2                    2

```

As you can see, only S factors. Please give an example of parameters that produce L factors:
```
$ biomarkers problem-create --bnet calzone_example.bnet --phenotype XXX
```

**Reply by Unaimend** [reply]
Ahh sry, I forgot to give you the phenotypes. 

```
biomarkers problem-create --problem calzone_problem.json --bnet calzone_example.bnet --phenotype "Survival,NonACD,Apoptosis" 
```

**Comment by hklarner** [comment, marked as answer]
Can you check whether this script works for you and does what you want?

```python
import pandas as pd

from biomarkers.tools.factorization import factorize_marker_sets
from biomarkers.tools.marker_detection import try_to_load_markers_or_exit
from biomarkers.tools.integer_sets import IntegerSets


if __name__ == "__main__":
    fname_factors_base = "tmp_factors.csv"      # <--- base name for export of long factors
    fname_markers = "tmp_markers.json"          # <--- input marker file
    size = 3                                    # <--- n_components of interest

    markers = try_to_load_markers_or_exit(fname=fname_markers)
    optimal_factors_by_size = factorize_marker_sets(markers=markers, fname_tex=None)
    component_names = markers.problem.component_names
    empty_row = tuple(0 for _ in component_names)
    factors = optimal_factors_by_size.get(size)

    if factors:
        long_factors = [integer_sets for integer_sets in factors if not integer_sets.is_singleton]

        if long_factors:
            for i, long_factor in enumerate(long_factors):
                long_factor: IntegerSets

                rows = []
                for s in long_factor._sets:
                    row = list(empty_row)
                    for j in s:
                        row[j] = 1

                    rows.append(row)

                df = pd.DataFrame(data=rows, columns=component_names)
                fname_factors = fname_factors_base.replace(".csv", f"_size{size}_{i:0>2}.csv")
                df.to_csv(fname_factors, index=False)
                print(f"created {fname_factors}")
```

**Reply by Unaimend** [reply]
Yes. Thank you very much. Thats exactly what I needed.
